### PR TITLE
Be more explicit about tank gauge delegate sizing

### PR DESCRIPTION
**Pull request by chriadam**
In TankGaugeGroup there is a Row populated by a Repeater whose delegate is a Loader.  The geometry of the Loader isn't explicitly defined, so will be inherited from its item if the Loader is active.  If a particular Loader is active, its geometry will be set; if that Loader then becomes inactive, it's possible that the geometry of that loader is being "stuck" at the old item's width, rather than being reset to zero.

This commit ensures that the width of the loader is explicitly set to the width of the item if active, otherwise zero.

It also makes various spacings and data bindings much more explicit for maintainability purposes.

A mock config is added which was intended to test the error case, although I was unable to reproduce using the mock backend.

Contributes to #1017

### CODE REVIEW

**Comment by chriadam** [comment]
I haven't verified that this fixes the issue.  I haven't been able to repro the original issue.

File: components/TankGaugeGroup.qml [hunk 1]
```diff
@@ -33,29 +34,41 @@ TankItem {
...
 
 		Repeater {
 			model: root.gaugeTanks
 			delegate: Loader {
-				active: model.index === 0 || root.mergeTanks
+				id: tankGaugeLoader
+				property int gaugeIndex: model.index
+				property real gaugeLevel: model.device.level
+				active: root.mergeTanks || gaugeIndex === root.delegateIndex
+				width: active && item ? item.width : 0
```
Comment (chriadam) [inline components/TankGaugeGroup.qml:46]: I think this line, width binding, is the real fix for the issue (although it's just a guess, since I can't repro the issue)

**Review by jpetrell — APPROVED** [review]
LGTM. These changes should fix the issue in reasonably high likelihood. I dislike the increased complexity a bit, would be nice to understand in detail what is going wrong, but we cannot really if we don't find a way to reproduce. Overall the way the model is used here in nested manner is a bit funky, usually you would update the model with right roles instead of constructing bunch of duplicated empty delegates. But it works, doesn't make sense to rewrite now.

FYI Getting empty gauges here also, but looks like the issue is unrelated but something that is already in main. Can be reproduced on desktop with the mock backend quite easily, usually seeing the empty gauges when going through configs 3.-6.

image

File: components/TankGaugeGroup.qml (hunk 1 above)
Comment (blammit) [reply components/TankGaugeGroup.qml:46]: Could you separate the bug-specific fixes into an individual commit?

File: components/TankGaugeGroup.qml [hunk 2]
```diff
@@ -33,29 +34,41 @@ TankItem {
...
+							// is some sort of race condition / property binding evaluation
+							// order issue which hits this codepath; see #1017
+							return 0
 						}
 					}
 
 					height: subgauges.height
 					valueType: root.tankProperties.valueType
-					value: (root.mergeTanks ? model.device.level : root.level) / 100
+					value: tankGaugeLoader.gaugeLevel / 100
```
Comment (blammit) [inline components/TankGaugeGroup.qml:68]: This is an unrelated change?

File: data/mock/TanksImpl.qml [hunk 3]
```diff
@@ -108,11 +109,45 @@ QtObject {
...
+		function onSetTanksFlaky(flaky) {
+			root.flakeTank.testingFlakyTanks = flaky
+		}
 	}
 
+	property Timer flakeTank: Timer {
+		running: Global.mockDataSimulator.timersActive && testingFlakyTanks
+		interval: 4 * 1000
+		repeat: true
+		property bool removeFlaky: false // whether to remove, or just invalidate, the flaky tank.
```
Comment (blammit) [inline data/mock/TanksImpl.qml:122]: Could you add some comments about what a "flaky" tank means?

File: components/TankGaugeGroup.qml [hunk 4]
```diff
@@ -33,29 +34,41 @@ TankItem {
...
 						const availableSpace = root.width - 2*Theme.geometry_levelsPage_subgauges_horizontalMargin
 						if (root.mergeTanks) {
 							return (availableSpace - (gaugeTanks.count - 1) * subgauges.spacing)/gaugeTanks.count
-						} else {
+						} else if (tankGaugeLoader.gaugeIndex === root.delegateIndex) {
 							return availableSpace
+						} else {
+							// shouldn't be necessary or useful, but just in case there
+							// is some sort of race condition / property binding evaluation
+							// order issue which hits this codepath; see #1017
```
Comment (blammit) [inline components/TankGaugeGroup.qml:61]: I'm not sure about adding this if we don't know whether it fixes the issue, and it creates more code paths we need to maintain.

File: pages/TanksTab.qml [hunk 5]
```diff
@@ -100,6 +100,7 @@ ListView {
 					totalRemaining: mergeTanks ? gaugeTanks.totalRemaining : model.device.remaining
 					totalCapacity: mergeTanks ? gaugeTanks.totalCapacity : model.device.capacity
 
+					delegateIndex: model.index
```
Comment (blammit) [inline pages/TanksTab.qml:103]: There are a few delegates floating around here, so might need a more specific name than `delegateIndex`

File: components/TankGaugeGroup.qml [hunk 6]
```diff
@@ -33,29 +34,41 @@ TankItem {
...
+		spacing: root.mergeTanks ? Theme.geometry_levelsPage_subgauges_spacing : 0
 
 		Repeater {
 			model: root.gaugeTanks
 			delegate: Loader {
-				active: model.index === 0 || root.mergeTanks
+				id: tankGaugeLoader
+				property int gaugeIndex: model.index
+				property real gaugeLevel: model.device.level
+				active: root.mergeTanks || gaugeIndex === root.delegateIndex
```
Comment (blammit) [inline components/TankGaugeGroup.qml:45]: There are a few delegates floating around here, so might need a more specific name than delegateIndex.

Comment (blammit) [reply components/TankGaugeGroup.qml:45]: What was this `active: root.mergeTanks || gaugeIndex === root.delegateIndex` change intended to do? The index of the tank within its group doesn't need to match the index of the group itself. These two indexes are from different models.

**Comment by blammit** [comment]
I'm seeing this with configuration 5 in mock mode:

image

**Comment by blammit** [comment]
> FYI Getting empty gauges here also, but looks like the issue is unrelated but something that is already in main. Can be reproduced on desktop with the mock backend quite easily, usually seeing the empty gauges when going through configs 3.-6.

I don't see this in main, only with this PR. What do you see @chriadam ?

**Comment by jpetrell** [comment]
> I don't see this in main, only with this PR. What do you see @chriadam ?

That may be thrue, not seeing it on main either now. Must have been confused with the branches.

**Comment by chriadam** [comment]
You're right, I could repro the issue you saw wiith my original PR.  Reduced to just the bugfix now.  I will investigate the cleanup next, but it can be separate.

**Comment by chriadam** [comment]
Hrm, still repro the issue with my smaller PR.  It might be the height binding change... Fix investigate and fix.

**Comment by chriadam** [comment]
Fixed.

File: data/mock/TanksImpl.qml (hunk 3 above)
Comment (chriadam) [reply data/mock/TanksImpl.qml]: Removed the "flaky" mock since it didn't reproduce the issue anyway.  Flaky just meant that it popped in and out of existence.  I was trying to reproduce the case in the report/issue, but couldn't.

File: components/TankGaugeGroup.qml (hunk 2 above)
Comment (chriadam) [reply components/TankGaugeGroup.qml]: Yes, it was cleanup.  That's now in #1138 instead.

File: components/TankGaugeGroup.qml (hunk 4 above)
Comment (chriadam) [reply components/TankGaugeGroup.qml]: Yes, it's not needed, removed.

File: components/TankGaugeGroup.qml (hunk 6 above)
Comment (chriadam) [reply components/TankGaugeGroup.qml]: I now use "groupIndex" as it's clearer.  And yes, the index of the tank now needs to match the index of the group, unless mergeTanks is true.  That is, I draw the "correct" tank rather than the zeroth tank.  Added comments etc.  But the changes are now in #1138 instead of this PR, as this PR now contains only the bugfix.

**Review by jpetrell — APPROVED** [review]

**Review by blammit — APPROVED** [review]

File: components/TankGaugeGroup.qml (hunk 6 above)
Comment (blammit) [reply components/TankGaugeGroup.qml]: Ah -- I see what you mean now, yep
